File: app/src/main/java/com/msk/minhascontas/listas/AdaptaListaPesquisaRC.java
package com.msk.minhascontas.listas;

import android.content.Context;
import android.content.res.Resources;
import android.database.Cursor;
import android.graphics.Color;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.msk.minhascontas.R;

import java.text.DateFormat;
import java.text.NumberFormat;
import java.util.Calendar;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

/**
 * Adapter para RecyclerView que exibe os resultados da pesquisa,
 * herdando a gestão do Cursor de CursorRecyclerViewAdapter.
 */
public class AdaptaListaPesquisaRC extends CursorRecyclerViewAdapter<AdaptaListaPesquisaRC.ViewHolder> {

    private final Resources res;
    private final String[] semana;
    private final String[] categoriaConta;
    private final NumberFormat dinheiro;
    private final DateFormat dataFormato;
    // Usamos um Set para IDs de contasListadas selecionadas para maior eficiência
    private final Set<Long> selecoes = new HashSet<>();

    // Constantes para cores (Carregadas via ContextCompat)
    private final int RED_COLOR;
    private final int GREEN_COLOR;
    private final int BLUE_COLOR;
    private final int SELECTION_COLOR;

    // Interface para manipular cliques (necessário no RecyclerView)
    private OnItemClickListener itemClickListener;

    public interface OnItemClickListener {
        void onItemClick(long id, int position);
        boolean onItemLongClick(long id, int position);
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.itemClickListener = listener;
    }

    // -------------------------------------------------------------------------
    // ViewHolder
    // -------------------------------------------------------------------------

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public final TextView nome;
        public final TextView categoria;
        public final TextView data;
        public final TextView dia;
        public final TextView valor;
        public final ImageView pagamento;

        public ViewHolder(View view) {
            super(view);
            // Referências de layout da linha_pesquisa
            nome = view.findViewById(R.id.tvNomeContaCriada);
            categoria = view.findViewById(R.id.tvNomeCategoria);
            data = view.findViewById(R.id.tvDataContaCriada);
            dia = view.findViewById(R.id.tvDiaContaCriada);
            valor = view.findViewById(R.id.tvValorContaCriada);
            pagamento = view.findViewById(R.id.ivPagamento);
        }
    }

    // -------------------------------------------------------------------------
    // Construtor
    // -------------------------------------------------------------------------

    public AdaptaListaPesquisaRC(Context context, Cursor cursor) {
        super(cursor);

        res = context.getResources();

        // 1. Inicialização das cores via ContextCompat
        RED_COLOR = ContextCompat.getColor(context, R.color.despesa_color);
        GREEN_COLOR = ContextCompat.getColor(context, R.color.aplicacao_color);
        BLUE_COLOR = ContextCompat.getColor(context, R.color.receita_color);
        SELECTION_COLOR = ContextCompat.getColor(context, R.color.linha_selecionada);

        // 2. Obtendo o Locale de forma moderna
        Locale current;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            current = res.getConfiguration().getLocales().get(0);
        } else {
            current = res.getConfiguration().locale;
        }

        dinheiro = NumberFormat.getCurrencyInstance(current);
        dataFormato = DateFormat.getDateInstance(DateFormat.SHORT, current);
        semana = res.getStringArray(R.array.Semana);
        categoriaConta = res.getStringArray(R.array.CategoriaConta);
    }

    // -------------------------------------------------------------------------
    // Implementação de RecyclerView.Adapter (Criação de View)
    // -------------------------------------------------------------------------

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        // Infla o layout da linha
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.linha_pesquisa, parent, false);

        final ViewHolder holder = new ViewHolder(view);

        // Configura os Listeners de clique
        view.setOnClickListener(v -> {
            if (itemClickListener != null) {
                // Notifica o clique com a ID do banco de dados e a posição
                itemClickListener.onItemClick(getItemId(holder.getAdapterPosition()), holder.getAdapterPosition());
            }
        });

        view.setOnLongClickListener(v -> {
            if (itemClickListener != null) {
                // Notifica o clique longo
                return itemClickListener.onItemLongClick(getItemId(holder.getAdapterPosition()), holder.getAdapterPosition());
            }
            return false;
        });

        return holder;
    }

    // -------------------------------------------------------------------------
    // Implementação de CursorRecyclerViewAdapter (Preenchimento de Dados)
    // -------------------------------------------------------------------------

    @Override
    public void onBindViewHolder(ViewHolder holder, Cursor cursor) {
        // Cursor já está na posição correta
        long idConta = getItemId(holder.getAdapterPosition());

        // Recuperar dados do Cursor
        String nomeconta = cursor.getString(1);
        int tipo = cursor.getInt(2);
        int classe = cursor.getInt(3);
        int categ = cursor.getInt(4);
        int d = cursor.getInt(5);
        int m = cursor.getInt(6);
        int a = cursor.getInt(7);
        String status = cursor.getString(9);
        int qtRepeticoes = cursor.getInt(10);
        int nrRepeticao = cursor.getInt(11);

        // Nome (com ajuste para recorrentes)
        holder.nome.setText(nomeconta);

        if (qtRepeticoes > 1 && (classe == 0 || classe == 3) && tipo == 0) {
            String nomeRecorrente = String.format(Locale.getDefault(), "%s %d/%d",
                    nomeconta, nrRepeticao, qtRepeticoes);
            holder.nome.setText(nomeRecorrente);
        }

        // Data e Dia da Semana
        Calendar c = Calendar.getInstance();
        c.set(a, m, d);
        int s = c.get(Calendar.DAY_OF_WEEK);
        holder.data.setText(dataFormato.format(c.getTime()));
        holder.dia.setText(semana[s - 1]);

        // Categoria e Classe
        String[] classeConta;
        if (tipo == 0) { // Despesa
            classeConta = res.getStringArray(R.array.TipoDespesa);
            holder.categoria.setText(String.format("%s | %s", classeConta[classe], categoriaConta[categ]));
        } else if (tipo == 1) { // Receita
            classeConta = res.getStringArray(R.array.TipoReceita);
            holder.categoria.setText(classeConta[classe]);
        } else { // Aplicação (tipo == 2)
            classeConta = res.getStringArray(R.array.TipoAplicacao);
            holder.categoria.setText(classeConta[classe]);
        }

        // Valor e Status de Pagamento/Cor
        holder.pagamento.setVisibility(View.INVISIBLE);
        holder.valor.setText(dinheiro.format(cursor.getDouble(8)));

        // Uso das constantes de cor
        if (tipo == 0) { // Despesa
            holder.valor.setTextColor(RED_COLOR);
            if ("paguei".equals(status))
                holder.pagamento.setVisibility(View.VISIBLE);
        } else if (tipo == 2) { // Aplicação
            holder.valor.setTextColor(GREEN_COLOR);
        } else { // Receita
            holder.valor.setTextColor(BLUE_COLOR);
            if ("paguei".equals(status))
                holder.pagamento.setVisibility(View.VISIBLE);
        }

        // Estado de Seleção
        if (selecoes.contains(idConta)) {
            holder.itemView.setBackgroundColor(SELECTION_COLOR);
        } else {
            holder.itemView.setBackgroundColor(Color.TRANSPARENT);
        }
    }

    // -------------------------------------------------------------------------
    // Métodos de Seleção
    // -------------------------------------------------------------------------

    /**
     * Marca/Desmarca uma conta pelo seu ID no banco de dados.
     * Atualiza apenas o item afetado (melhor performance).
     */
    public void marcaConta(long idConta, Boolean seleciona) {
        int posicao = -1;

        // Localiza a posição do item com o ID dado
        Cursor cursor = getCursor();
        if (cursor != null && cursor.moveToFirst()) {
            int idColumnIndex = cursor.getColumnIndexOrThrow("_id");
            for (int i = 0; i < cursor.getCount(); i++) {
                cursor.moveToPosition(i);
                if (cursor.getLong(idColumnIndex) == idConta) {
                    posicao = i;
                    break;
                }
            }
        }

        if (seleciona) {
            selecoes.add(idConta);
        } else {
            selecoes.remove(idConta);
        }

        // Se a posição foi encontrada, atualiza apenas aquele item
        if (posicao != -1) {
            notifyItemChanged(posicao);
        }
    }

    /**
     * Limpa todas as seleções.
     */
    public void limpaSelecao() {
        if (!selecoes.isEmpty()) {
            selecoes.clear();
            // Notifica que toda a lista mudou para resetar os fundos
            notifyDataSetChanged();
        }
    }

    /**
     * Retorna o Set de IDs de contasListadas selecionadas.
     */
    public Set<Long> getSelecoes() {
        return selecoes;
    }
}

File: app/src/main/java/com/msk/minhascontas/features/listas/AdaptaListaPesquisaRC.kt
package com.msk.minhascontas.features.listas

import android.content.Context
import android.content.res.Resources
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.msk.minhascontas.R
import com.msk.minhascontas.db.Conta
import com.msk.minhascontas.utils.LabelUtils
import java.text.DateFormat
import java.text.NumberFormat
import java.util.Calendar
import java.util.Locale

/**
 * Adapter para RecyclerView que exibe os resultados da pesquisa usando List<Conta>.
 * Migrado de Cursor para uma implementação mais segura com Room.
 */
class AdaptaListaPesquisaRC(context: Context, private var contas: List<Conta> = emptyList()) :
    RecyclerView.Adapter<AdaptaListaPesquisaRC.ViewHolder>() {
    private val res: Resources = context.resources
    private val semana: Array<String?>
    private val categoriaConta: Array<String?>
    private val dinheiro: NumberFormat
    private val dataFormato: DateFormat

    /**
     * Retorna o Set de IDs de contas selecionadas.
     */
    val selecoes: MutableSet<Long> = HashSet()

    // Constantes para cores
    private val redColor: Int = ContextCompat.getColor(context, R.color.despesa_color)
    private val greenColor: Int = ContextCompat.getColor(context, R.color.aplicacao_color)
    private val blueColor: Int = ContextCompat.getColor(context, R.color.receita_color)
    private val selectionColor: Int = ContextCompat.getColor(context, R.color.linha_selecionada)

    private var itemClickListener: OnItemClickListener? = null

    interface OnItemClickListener {
        fun onItemClick(id: Long, position: Int)
        fun onItemLongClick(id: Long, position: Int): Boolean
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        this.itemClickListener = listener
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val nome: TextView = view.findViewById(R.id.tvNomeContaCriada)
        val categoria: TextView = view.findViewById(R.id.tvNomeCategoria)
        val data: TextView = view.findViewById(R.id.tvDataContaCriada)
        val dia: TextView = view.findViewById(R.id.tvDiaContaCriada)
        val valor: TextView = view.findViewById(R.id.tvValorContaCriada)
        val pagamento: ImageView = view.findViewById(R.id.ivPagamento)
    }

    init {
        val current: Locale = res.configuration.locales.get(0)
        dinheiro = NumberFormat.getCurrencyInstance(current)
        dataFormato = DateFormat.getDateInstance(DateFormat.SHORT, current)
        semana = res.getStringArray(R.array.Semana)
        categoriaConta = res.getStringArray(R.array.CategoriaConta)
        setHasStableIds(true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view: View = LayoutInflater.from(parent.context)
            .inflate(R.layout.linha_pesquisa, parent, false)
        val holder = ViewHolder(view)

        view.setOnClickListener {
            val pos = holder.bindingAdapterPosition
            if (pos != RecyclerView.NO_POSITION) {
                itemClickListener?.onItemClick(getItemId(pos), pos)
            }
        }

        view.setOnLongClickListener {
            val pos = holder.bindingAdapterPosition
            if (pos != RecyclerView.NO_POSITION) {
                itemClickListener?.onItemLongClick(getItemId(pos), pos) ?: false
            } else false
        }

        return holder
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val conta = contas[position]
        val idConta = conta.idConta

        holder.nome.text = conta.nome
        if (conta.qtRepete > 1 && (conta.classeConta == 0 || conta.classeConta == 3) && conta.tipo == 0) {
            holder.nome.text = String.format(Locale.getDefault(), "%s %d/%d", conta.nome, conta.nRepete, conta.qtRepete)
        }

        val c = Calendar.getInstance()
        c.set(conta.ano, conta.mes - 1, conta.dia)
        holder.data.text = dataFormato.format(c.time)
        holder.dia.text = semana[c.get(Calendar.DAY_OF_WEEK) - 1]

        holder.categoria.text = LabelUtils.getClasseLabel(holder.itemView.context, conta.tipo, conta.classeConta)
        if (conta.tipo == 0) {
            holder.categoria.text = "${holder.categoria.text} | ${categoriaConta[conta.categoria]}"
        }

        holder.valor.text = dinheiro.format(conta.valor)
        holder.pagamento.visibility = View.INVISIBLE

        when (conta.tipo) {
            0 -> {
                holder.valor.setTextColor(redColor)
                if ("paguei" == conta.pagamento) holder.pagamento.visibility = View.VISIBLE
            }
            2 -> holder.valor.setTextColor(greenColor)
            else -> {
                holder.valor.setTextColor(blueColor)
                if ("paguei" == conta.pagamento) holder.pagamento.visibility = View.VISIBLE
            }
        }

        holder.itemView.setBackgroundColor(if (selecoes.contains(idConta)) selectionColor else Color.TRANSPARENT)
    }

    override fun getItemCount(): Int = contas.size
    override fun getItemId(position: Int): Long = contas[position].idConta

    fun swapList(newList: List<Conta>) {
        this.contas = newList
        notifyDataSetChanged()
    }

    fun marcaConta(idConta: Long, seleciona: Boolean) {
        val position = contas.indexOfFirst { it.idConta == idConta }
        if (seleciona) selecoes.add(idConta) else selecoes.remove(idConta)
        if (position != -1) notifyItemChanged(position)
    }

    fun limpaSelecao() {
        if (selecoes.isNotEmpty()) {
            selecoes.clear()
            notifyDataSetChanged()
        }
    }
}
